### ISSUE TITLE
as_coco function not saving the annotations as expected

### ISSUE THREAD

**Issue by smarabattula** [roboflow/supervision#383]
### Search before asking

- [X] I have searched the Supervision [issues](https://github.com/roboflow/supervision/issues) and found no similar bug report.

### Bug

Hi

I'm working with an object detection dataset in RoboFlow and getting segmentation masks to apply image segmentation. I'm trying to modify the dataset's annotations by applying SAM model and a custom function (most of the code has been written from RoboFlow notebook snippets). However, when I tried to save the coco annotations file using as_coco function, it isn't saving for me as expected. I tried testing it like this: after as_coco, use from_coco and compare the Detectiondataset objects before as_coco and after as_coco. It turned out to be False.

I went through supervision community posts but couldn't find any ones related to my issue.
Can you please verify if this is a bug? Or am I missing something here? 

Attached code snippets and output result for reference.
![image](https://github.com/roboflow/supervision/assets/122768312/96e277b1-0407-46fe-9c59-cf14c049c8e1)
![image](https://github.com/roboflow/supervision/assets/122768312/c38c65c1-6dbf-43d3-a7b6-942b50f85921)

### Environment

- Supervision: 0.14.0
- VS Code
- Windows 11
- Python: 3.11.1

### Minimal Reproducible Example

            object_detection_dataset = sv.DetectionDataset.from_coco(
                images_directory_path=IMAGES_DIRECTORY_PATH,
                annotations_path=ANNOTATIONS_FILE_PATH
            )

            image_names = list(object_detection_dataset.images.keys()) 

            for image_name in image_names:
                    time_img_start = time.time()
                    # Process the image and annotations here
                    # detection = detections[image_name]
                    object_detection_dataset.annotations[image_name].mask = segment(sam_predictor=sam_predictor, image=cv2.imread(image_name, cv2.COLOR_BGR2RGB),xyxy=object_detection_dataset.annotations[image_name].xyxy)
                    time_img_end = time.time()
                    image_name = image_name.replace(IMAGES_DIRECTORY_PATH + "\\", "")
                    print(image_name + " processing time: " + str(time_img_end - time_img_start) + " secs")
                    print(sys.getsizeof(object_detection_dataset))
                    times[DATA_SET_SUBDIRECTORY][image_name] = time_img_end - time_img_start
                    break
            object_detection_dataset.as_coco(annotations_path=ANNOTATIONS_FILE_PATH, images_directory_path=None)
            object_detection_dataset1 = sv.DetectionDataset.from_coco(
                images_directory_path=IMAGES_DIRECTORY_PATH,
                annotations_path=ANNOTATIONS_FILE_PATH
            )

            print("************************************************************")
            print(object_detection_dataset == object_detection_dataset1)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by smarabattula** [comment]
Upon checking, I've seen that if I use from_coco function again, it reads the unchanged version of annotation file.

**Comment by SkalskiP** [comment]
Hi @smarabattula 👋🏻 

Taking this into consideration:

> Upon checking, I've seen that if I use from_coco function again, it reads the unchanged version of annotation file.

Do you still experience any problems?

**Comment by smarabattula** [comment]
@SkalskiP Yes, the issue still seems to be existing. The function as_coco isn't saving the updated annotations (but the changes are visible in VS Code). So when I try to read the file again using from_coco, it still reads the unedited version of it. the issue with as_coco function still exists.

**Comment by SkalskiP** [comment]
Hi @smarabattula 👋🏻 Understood. I'll try my best to help you out. I'm going to need a minimal example that I could run (preferably in Google Colab) that I could use to understand the problem.

I know you already gave us some code, but it uses variables and functions not defined in your snippet. We would need something simple.

**Comment by smarabattula** [comment]
Here is a running code

import random
import cv2
import numpy as np

import supervision as sv
import os
from segment_anything import sam_model_registry,SamPredictor
import torch

DEVICE = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
MODEL_TYPE = "vit_h"

HOME = os.getcwd()

# Download SAM Weights into weights directory
CHECKPOINT_PATH = os.path.join(HOME, "weights", "sam_vit_h_4b8939.pth")

sam = sam_model_registry[MODEL_TYPE](checkpoint=CHECKPOINT_PATH).to(device=DEVICE)

sam_predictor = SamPredictor(sam)

mask_annotator = sv.MaskAnnotator()
box_annotator = sv.BoxAnnotator()

random.seed(2023)

# Segmentation algorithm function, as given in Roboflow Notebook
def segment(sam_predictor: SamPredictor, image: np.ndarray, xyxy: np.ndarray) -> np.ndarray:
    sam_predictor.set_image(image)
    result_masks = []
    for box in xyxy:
        masks, scores, logits = sam_predictor.predict(
            box=box,
            multimask_output=False
        )
        index = np.argmax(scores)
        result_masks.append(masks[index])
    return np.array(result_masks)

# Roboflow loading downloaded dataset from local 
roboflow.login()
rf = Roboflow()
project = rf.workspace("Workspace_name").project("project_name") #Workspace_name , project_name masked
dataset = project.version(20) # Already downloaded the dataset

# Paths definition
dataset_loc = os.path.join(HOME,dataset.name+"-"+dataset.version)
ANNOTATIONS_FILE_NAME = '_annotations.coco.json'
DATA_SET_SUBDIRECTORY = 'test'
IMAGES_DIRECTORY_PATH = os.path.join(dataset_loc, DATA_SET_SUBDIRECTORY) 
ANNOTATIONS_FILE_PATH = os.path.join(IMAGES_DIRECTORY_PATH, ANNOTATIONS_FILE_NAME)

# Loading the dataset using from_coco
object_detection_dataset = sv.DetectionDataset.from_coco(
    images_directory_path=IMAGES_DIRECTORY_PATH,
    annotations_path=ANNOTATIONS_FILE_PATH
)

# Getting a random image name
image_names = list(object_detection_dataset.images.keys()) 
image_name = random.choice(image_names)

# Applying segment function to update image annotation's mask attribute
object_detection_dataset.annotations[image_name].mask = segment(
sam_predictor=sam_predictor, image=cv2.imread(image_name, cv2.COLOR_BGR2RGB),xyxy=object_detection_dataset.annotations[image_name].xyxy)

# Saving annotation using as_coco (ISSUE HERE)
object_detection_dataset.as_coco(annotations_path=ANNOTATIONS_FILE_PATH, images_directory_path=None)

# Reading the dataset again to compare 
object_detection_dataset1 = sv.DetectionDataset.from_coco(
    images_directory_path=IMAGES_DIRECTORY_PATH,
    annotations_path=ANNOTATIONS_FILE_PATH
)

# Output is FALSE!
print(object_detection_dataset == object_detection_dataset1)

**Comment by smarabattula** [comment]
Hey @SkalskiP, would like to know if you are getting similar issues or if this is just an issue on my end. 
By the way, I looked into the code and tried modifying the underlying function for from_coco and as_coco. The save_json_file looks to be saving properly, I opened the annotations file to check and annotation got updated. But for some reason the read_json_file reads an older copy of the annotations file, leading to this issue. Could it be due to caching?

**Comment by SkalskiP** [comment]
Hi, @smarabattula! 👋🏻 Welcome back after the weekend.

I don't think it is a caching issue. There is no caching:) Could you prepare a google colab that we could use for collaborative debugging?

**Comment by SkalskiP** [comment]
Hi @smarabattula 👋🏻, I'll try to take a look at it as fast as possible.

**Comment by smarabattula** [comment]
Hey @SkalskiP 🤝 ,
Did you get a chance to look at the code?

**Comment by SkalskiP** [comment]
@smarabattula I'm really sorry but I was extremely busy with new release last week.

**Comment by smarabattula** [comment]
Hi @SkalskiP,

Following up on the issue. Any update?

Thanks

**Comment by SkalskiP** [comment]
Hi @smarabattula 👋🏻 

Unfortunately, not yet. :/
